server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapStarTreeBuilder.java

```diff
@@ -66,34 +62,136 @@ public List<StarTreeDocument> getStarTreeDocuments() {
+     */
+    StarTreeDocument[] mergeStarTreeValues(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        List<StarTreeDocument> starTreeDocuments = new ArrayList<>();
+        for (StarTreeValues starTreeValues : starTreeValuesSubs) {
+            List<Dimension> dimensionsSplitOrder = starTreeValues.getStarTreeField().getDimensionsOrder();
+            SequentialDocValuesIterator[] dimensionReaders = new SequentialDocValuesIterator[starTreeValues.getStarTreeField()
+                .getDimensionsOrder()
+                .size()];
+
+            for (int i = 0; i < dimensionsSplitOrder.size(); i++) {
+                String dimension = dimensionsSplitOrder.get(i).getField();
+                dimensionReaders[i] = new SequentialDocValuesIterator(starTreeValues.getDimensionDocValuesIteratorMap().get(dimension));
+            }
+
+            List<SequentialDocValuesIterator> metricReaders = new ArrayList<>();
+            for (Map.Entry<String, DocIdSetIterator> metricDocValuesEntry : starTreeValues.getMetricDocValuesIteratorMap().entrySet()) {
+                metricReaders.add(new SequentialDocValuesIterator(metricDocValuesEntry.getValue()));
+            }
+
+            boolean endOfDoc = false;
+            int currentDocId = 0;
+            int numSegmentDocs = Integer.parseInt(
+                starTreeValues.getAttributes().getOrDefault(NUM_SEGMENT_DOCS, String.valueOf(DocIdSetIterator.NO_MORE_DOCS))
+            );
+            while (!endOfDoc) {
+                Long[] dims = new Long[starTreeValues.getStarTreeField().getDimensionsOrder().size()];
+                int i = 0;
+                int endOfDocCounter = 0;
+                for (SequentialDocValuesIterator dimensionDocValueIterator : dimensionReaders) {
+                    int doc = dimensionDocValueIterator.nextDoc(currentDocId);
+                    Long val = dimensionDocValueIterator.value(currentDocId);
+                    if (doc == DocIdSetIterator.NO_MORE_DOCS) {
+                        endOfDocCounter++;
+                    }
+                    dims[i] = val;
+                    i++;
+                }
+                // we've exhausted all dimension readers
+                if (endOfDocCounter == dims.length) {
+                    break;
+                }
+                i = 0;
+                Object[] metrics = new Object[metricReaders.size()];
+                for (SequentialDocValuesIterator metricDocValuesIterator : metricReaders) {
+                    metricDocValuesIterator.nextDoc(currentDocId);
+                    // As part of merge, we traverse the star tree doc values
+                    // The type of data stored in metric fields is different from the
+                    // actual indexing field they're based on
+                    metrics[i] = metricAggregatorInfos.get(i)
+                        .getValueAggregators()
+                        .toStarTreeNumericTypeValue(metricDocValuesIterator.value(currentDocId));
```

Comment:
Lets ensure this handles the identity value for cases when metric stat is not present for a segment

Reply:
For now initialized to 0.0 if value is null for sum aggregator.

Reply:
This is being solved as part of all aggregators.